kotlin/samples/containers/hello-back-button/src/main/java/com/squareup/sample/hellobackbutton/HelloBackButtonWorkflow.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sample.hellobackbutton
+
+import com.squareup.sample.hellobackbutton.HelloBackButtonWorkflow.Rendering
+import com.squareup.sample.hellobackbutton.HelloBackButtonWorkflow.State
+import com.squareup.sample.hellobackbutton.HelloBackButtonWorkflow.State.Able
+import com.squareup.sample.hellobackbutton.HelloBackButtonWorkflow.State.Baker
+import com.squareup.sample.hellobackbutton.HelloBackButtonWorkflow.State.Charlie
+import com.squareup.workflow.RenderContext
+import com.squareup.workflow.Snapshot
+import com.squareup.workflow.StatefulWorkflow
+import com.squareup.workflow.action
+
+object HelloBackButtonWorkflow : StatefulWorkflow<Unit, State, Nothing, Rendering>() {
+  enum class State {
+    Able,
+    Baker,
+    Charlie;
+  }
+
+  data class Rendering(
+    val message: String,
+    val onClick: () -> Unit,
+    val onBackPressed: (() -> Unit)?
+  )
+
+  override fun initialState(
+    props: Unit,
+    snapshot: Snapshot?
+  ): State = Able
+
+  override fun render(
+    props: Unit,
+    state: State,
+    context: RenderContext<State, Nothing>
+  ): Rendering {
+    return Rendering(
+        message = "$state",
+        onClick = { context.actionSink.send(advance) },
+        onBackPressed = { context.actionSink.send(retreat) }.takeIf { state != Able }
```

Comment:
This forces a lambda allocation even if it's not used - maybe better for sample code to use a regular if?

Reply:
It's actually tricky to keep that readable, which is why I cheated and went with `takeIf` -- justified it to myself because this is sample code. :)

```kotlin
{
    val onBack: (() -> Unit)? = if (state == Able) {
      null
    } else {
      { context.actionSink.send(retreat) }
    }

    return Rendering(
        message = "$state",
        onClick = { context.actionSink.send(advance) },
        onBackPressed = onBack
    )
  }
```

WDYT?